iosHyperskillApp/iosHyperskillApp/Sources/Modules/StudyPlan/Views/StudyPlanSectionItemView.swift

```diff
@@ -0,0 +1,182 @@
+                            .foregroundColor(appearance.progressBadgeForegroundColor)
+                            .padding(appearance.progressBadgeInsets.edgeInsets)
+                            .background(appearance.progressBadgeBackgroundColor)
+                            .cornerRadius(appearance.progressBadgeCornerRadius)
+                    }
+                }
+
+                Spacer()
+
+                buildItemIcon(itemState: item.state)
+            }
+            .padding()
+            .frame(minHeight: appearance.buttonStyle.minHeight)
+            .background(
+                buildCompletenessBar(
+                    completenessPercentage: Float(truncating: item.progress ?? 0)
+                )
+            )
+        }
+        .disabled(!item.isClickable)
+        .buttonStyle(appearance.buttonStyle)
+        .buttonStyle(BounceButtonStyle())
+    }
+
+    // MARK: Private API
+
+    @ViewBuilder
+    private func buildCompletenessBar(completenessPercentage: Float) -> some View {
+        GeometryReader { geometry in
+            Rectangle()
+                .stroke(lineWidth: 0)
+                .background(appearance.backgroundProgressColor)
+                .cornerRadius(appearance.buttonStyle.cornerRadius)
+                .frame(
+                    width: geometry.size.width * CGFloat(completenessPercentage) / 100
+                )
+        }
+    }
+
+    @ViewBuilder
+    private func buildItemIcon(
+        itemState: StudyPlanWidgetViewStateSectionItemState
+    ) -> some View {
+        switch itemState {
+        case StudyPlanWidgetViewStateSectionItemState.completed:
+            Image(systemName: "checkmark")
+                .resizable()
+                .renderingMode(.template)
+                .aspectRatio(contentMode: .fit)
+                .foregroundColor(Color(ColorPalette.secondary))
+                .frame(widthHeight: appearance.rightDetailIconWidthHeight)
+        case StudyPlanWidgetViewStateSectionItemState.locked:
+            Image(systemName: "lock")
+                .resizable()
+                .renderingMode(.template)
+                .aspectRatio(contentMode: .fit)
+                .foregroundColor(Color(ColorPalette.onSurfaceAlpha38))
+                .frame(widthHeight: appearance.rightDetailIconWidthHeight)
+        case StudyPlanWidgetViewStateSectionItemState.next:
+            Image(Images.Home.ProblemOfDay.arrowUncompleted)
+                .renderingMode(.original)
+                .resizable()
+                .frame(widthHeight: appearance.arrowIconSize)
+        case StudyPlanWidgetViewStateSectionItemState.skipped:
+            Image(Images.Track.TopicsToDiscoverNext.skippedTopic)
+                .resizable()
+                .renderingMode(.template)
+                .aspectRatio(contentMode: .fit)
+                .foregroundColor(Color(ColorPalette.onSurfaceAlpha38))
+                .frame(widthHeight: appearance.rightDetailIconWidthHeight)
```

Comment:
Code duplication.

You can create `StudyPlanSectionItemIconView` that will build an `Image` based on `StudyPlanWidgetViewStateSectionItemState`, you can implement some computed properties (`renderingMode`, `foregroundColor `, ...)

MOKO KSwift not generated a wrapper around `StudyPlanWidgetViewStateSectionItemState`? If so we can add it manually, with wrapper it will be much simpler to implement `StudyPlanSectionItemIconView`.

Reply:
StudyPlanWidgetViewStateSectionItemState is enum, so MOKO KSwift doesn't generate wrapper

Reply:
@vladkash moko-kswift has an issue about this "[Generate Swift enum class for Kotlin enum](https://github.com/icerockdev/moko-kswift/issues/7)" and actually there is [implementation of this feature](https://github.com/cl3m/kmp-redux/blob/develop/buildSrc/src/main/kotlin/EnumGenerator.kt). We can wait until the author of this feature submits a PR to the moko-kswift or we can [integrate it via the generator](https://github.com/icerockdev/moko-kswift#implementation-of-own-generator).